addons/account/tests/test_reconciliation_matching_rules.py

```diff
@@ -347,66 +387,377 @@ def test_matching_fields_match_label(self):
-        self.bank_line_1.amount += 10
+    @freeze_time('2019-01-01')
+    def test_zero_payment_tolerance_outbound_auto_reconcile(self):
+        rule = self._create_rule(
+            auto_reconcile=True,
+            line_ids=[{}],
+        )
+
+        invl = self._create_invoice_line(1000.0, self.partner_a, 'in_invoice',  pay_reference='123456', date='2019-01-01')
+
+        # The payment amount is higher than the invoice one.
+        st_line = self._create_st_line(amount=-1010.0, payment_ref='123456')
+        self._check_statement_matching(
+            rule,
+            {st_line.id: {'aml_ids': invl.ids, 'model': rule, 'partner': st_line.partner_id, 'status': 'reconciled'}},
+            statements=st_line.statement_id,
+        )
+
+        self.assertRecordValues(st_line.line_ids, [
+            {'debit': 0.0,      'credit': 1010.0,   'account_id': self.bank_journal.default_account_id.id,  'reconciled': False},
+            {'debit': 10.0,     'credit': 0.0,      'account_id': invl.account_id.id,                       'reconciled': False},
+            {'debit': 1000.0,   'credit': 0.0,      'account_id': invl.account_id.id,                       'reconciled': True},
+        ])
 
-        # Check match_total_amount: line amount >= total residual amount, match_total_amount_param just not matched.
-        self.rule_1.match_total_amount_param = 90.0
-        self.bank_line_1.amount += 10.01
-        self._check_statement_matching(self.rule_1, {
-            self.bank_line_1.id: {'aml_ids': []},
-            self.bank_line_2.id: {'aml_ids': [
-                self.invoice_line_1.id,
-                self.invoice_line_2.id,
-                self.invoice_line_3.id,
-            ], 'model': self.rule_1, 'partner': self.bank_line_2.partner_id},
-            self.cash_line_1.id: {'aml_ids': [self.invoice_line_4.id], 'model': self.rule_1, 'partner': self.cash_line_1.partner_id},
-        })
-        self.rule_1.match_total_amount_param = 100.0
-        self.bank_line_1.amount -= 10.01
+    @freeze_time('2019-01-01')
+    def test_not_enough_payment_tolerance_inbound(self):
+        rule = self._create_rule(
+            payment_tolerance_param=0.5,
+            line_ids=[{}],
+        )
+
+        invl = self._create_invoice_line(1000.0, self.partner_a, 'out_invoice', date='2019-01-01')
+
+        # No matching because there is no enough tolerance.
+        st_line = self._create_st_line(amount=990.0)
+        self._check_statement_matching(
+            rule,
+            {st_line.id: {'aml_ids': []}},
+            statements=st_line.statement_id,
+        )
+
+        # The payment amount is higher than the invoice one. Tolerance is not checked.
+        st_line = self._create_st_line(amount=1010.0)
+        self._check_statement_matching(
+            rule,
+            {st_line.id: {'aml_ids': invl.ids, 'model': rule, 'partner': st_line.partner_id}},
+            statements=st_line.statement_id,
+        )
+
```

Comment:
Duplicating the tests between inbound and outbound is a pity ; why not writing a loop each time, wich then takes care of each type ?